File: server/models/barCodes.ts
import { type Database } from "sqlite3"

export class BarCodesModel {
  @Library('db') private db: Database
  public async getAll(): Promise<Miscellaneous.BarCode[]> {
    const results = await new Promise<Miscellaneous.BarCodeResult[]>(resolve => {
      this.db.all<Miscellaneous.BarCodeResult>('SELECT rowid, * FROM bar_codes', (err, rows) => {
        if (err) {
          console.error(err)
          resolve([])
        } else {
          resolve(rows)
        }
      })
    })
    const barCodes: Miscellaneous.BarCode[] = []
    for (const { rowid, name, tag, value } of results) {
      barCodes.push({
        id: rowid,
        name,
        tag,
        value
      })
    }
    return barCodes
  }
  public async get(id: Miscellaneous.BarCode['id']): Promise<Miscellaneous.BarCode | null> {
    if (isNaN(id)) {
      return null
    }
    return new Promise<Miscellaneous.BarCode | null>(resolve => {
      this.db.get<Miscellaneous.BarCodeResult>('SELECT rowid, * FROM bar_codes WHERE rowid = ?', [id], (err, row) => {
        if (err) {
          console.log(err)
          resolve(null)
        } else {
          if (row) {
            resolve({
              id: row.rowid,
              name: row.name,
              tag: row.tag,
              value: row.value
            })
          } else {
            resolve(null)
          }
        }
      })
    })
  }
  public async create({ name, tag, value }: Miscellaneous.NewBarCode): Promise<void> {
    await new Promise<void>(resolve => {
      this.db.run('INSERT INTO bar_codes (name, tag, value) VALUES (?, ?, ?)', [name, tag || '', value], (err) => {
        if (err) {
          console.log(err)
        }
        resolve()
      })
    })
  }
  public async update(id: Miscellaneous.BarCode['id'], { name, tag, value }: Miscellaneous.NewBarCode): Promise<void> {
    if (isNaN(id)) {
      return
    }
    await new Promise<void>(resolve => {
      this.db.run('UPDATE bar_codes SET name = ?, tag = ?, value = ? WHERE rowid = ?', [name, tag || '', value, id], (err) => {
        if (err) {
          console.log(err)
        }
        resolve()
      })
    })
  }
  public async delete(id: Miscellaneous.BarCode['id']): Promise<void> {
    if (isNaN(id)) {
      return
    }
    await new Promise<void>(resolve => {
      this.db.run('DELETE FROM bar_codes WHERE rowid = ?', [id], (err) => {
        if (err) {
          console.log(err)
        }
        resolve()
      })
    })
  }
}

File: drivers/sqlite/servers/barcodes.go
package servers

import (
	"Miscellaneous/plugins/grpcs"
	"Miscellaneous/sqlite/db"
	"context"
)

type BarCodesServer struct {
	grpcs.UnimplementedBarCodesServer
}

func (bcm BarCodesServer) Create(ctx context.Context, in *grpcs.NewBarCode) (*grpcs.Empty, error) {
	_, err := db.Client.Exec(
		"INSERT INTO bar_codes (name, tag, value) VALUES (?, ?, ?)",
		in.GetName(), in.GetTag(), in.GetValue(),
	)
	if err != nil {
		return nil, err
	}
	return &grpcs.Empty{}, nil
}

func (bcm BarCodesServer) Get(ctx context.Context, in *grpcs.IdRequest) (*grpcs.BarCode, error) {
	var result grpcs.BarCode

	err := db.Client.QueryRow(
		"SELECT rowid, * FROM bar_codes WHERE rowid = ?",
		in.GetId(),
	).Scan(&result.Id, &result.Name, &result.Tag, &result.Value)
	if err != nil {
		if err.Error() == "sql: no rows in result set" {
			return nil, nil
		}
		return nil, err
	}

	return &result, nil
}

func (bcm BarCodesServer) GetAll(ctx context.Context, _ *grpcs.Empty) (*grpcs.BarCodeList, error) {
	rows, err := db.Client.Query("SELECT rowid, * FROM bar_codes")
	if err != nil {
		return nil, err
	}

	results := []*grpcs.BarCode{}
	result := &grpcs.BarCodeList{Barcodes: results}

	for rows.Next() {
		var row grpcs.BarCode
		err := rows.Scan(&row.Id, &row.Name, &row.Tag, &row.Value)
		if err != nil {
			continue
		}

		result.Barcodes = append(result.Barcodes, &row)
	}

	return result, nil
}

func (bcm BarCodesServer) Update(ctx context.Context, in *grpcs.BarCode) (*grpcs.Empty, error) {
	_, err := db.Client.Exec(
		"UPDATE bar_codes SET name = ?, tag = ?, value = ? WHERE rowid = ?",
		in.GetName(), in.GetTag(), in.GetValue(), in.GetId(),
	)
	if err != nil {
		return nil, err
	}
	return &grpcs.Empty{}, nil
}

func (bcm BarCodesServer) Delete(ctx context.Context, in *grpcs.IdRequest) (*grpcs.Empty, error) {
	_, err := db.Client.Exec("DELETE FROM bar_codes WHERE rowid = ?", in.GetId())
	if err != nil {
		return nil, err
	}
	return &grpcs.Empty{}, nil
}
